### ISSUE TITLE
`const [y, y] = []` throws a TypeError, instead of SyntaxError

### ISSUE THREAD

**Issue by p-bakker** [mozilla/rhino#989]
Followup of https://github.com/mozilla/rhino/issues/386

**Comment by tuchida** [comment]
Perhaps this is the same problem.
```js
const y = 1;
const y = 2;
```
It should not be possible to achieve this without implementing TDZ.

**Comment by p-bakker** [comment]
Is this really related to TDZ? IMHO this is just about redeclaration of a `const`, whereas TDZ is about accessing a `const` before it is declared, with let/const not being hoisted to the top of a function like vars, but only being defined as the point in the code where they are declared, so any code accessing them before they are declared is what TDZ is about (AFAIK)

**Comment by tuchida** [comment]
Sorry, this may not be directly related.
However, if the redeclaration check is to be done in the parse phase instead of the execution phase, I think TDZ will have to be implemented at the same time.

**Comment by p-bakker** [comment]
To my knowledge TDZ issues throw a ReferenceError and I think those are runtime errors, not parse-time errors.

And redeclaration of const/let throws a SyntaxError, which is a parse-time error I think

**Comment by tuchida** [comment]
After applying these, I got the following error.
```
$ ./gradlew run -q --console=plain --args="-version 200"
Rhino 1.7.16-SNAPSHOT
js> const [y, y] = []
js: line 2: TypeError: redeclaration of const y.
js: 
js: ^
java.lang.IllegalStateException: FAILED ASSERTION
	at org.mozilla.javascript.Kit.codeBug(Kit.java:356)
	at org.mozilla.javascript.IRFactory.transformScript(IRFactory.java:950)
	at org.mozilla.javascript.IRFactory.transform(IRFactory.java:198)
	at org.mozilla.javascript.IRFactory.transformTree(IRFactory.java:128)
	at org.mozilla.javascript.Context.parse(Context.java:2526)
	at org.mozilla.javascript.Context.compileImpl(Context.java:2443)
	at org.mozilla.javascript.Context.compileString(Context.java:1405)
	at org.mozilla.javascript.Context.compileString(Context.java:1393)
	at org.mozilla.javascript.tools.shell.Main.processSource(Main.java:480)
	at org.mozilla.javascript.tools.shell.Main.processFiles(Main.java:180)
	at org.mozilla.javascript.tools.shell.Main$IProxy.run(Main.java:98)
	at org.mozilla.javascript.Context.call(Context.java:549)
	at org.mozilla.javascript.ContextFactory.call(ContextFactory.java:475)
	at org.mozilla.javascript.tools.shell.Main.exec(Main.java:164)
	at org.mozilla.javascript.tools.shell.Main.main(Main.java:142
```

**Comment by tuchida** [comment]
Errors whose message starts with `TypeError: ` are `TypeError`.
https://github.com/mozilla/rhino/blob/7feaad0e0806e48728e4c7754654b8146db38c30/rhino/src/main/java/org/mozilla/javascript/DefaultErrorReporter.java#L39-L55

Perhaps all of these are fine with `SyntaxError`.
https://github.com/mozilla/rhino/blob/7feaad0e0806e48728e4c7754654b8146db38c30/rhino/src/main/resources/org/mozilla/javascript/resources/Messages.properties#L268-L281
https://github.com/mozilla/rhino/blob/7feaad0e0806e48728e4c7754654b8146db38c30/rhino/src/main/resources/org/mozilla/javascript/resources/Messages.properties#L474-L478
```js
// old generator
function g() { return 0; yield 1; }
// js: "<stdin>", line 2: TypeError: generator function g returns a value
// js: function g() { return 0; yield 1; }
// js: ................................^
// js: "<stdin>", line 2: Compilation produced 1 syntax errors.

// latest generator
function *g() { return 0; yield 1; }  // no error
```

However, `TypeError` is correct if the constant is overwritten at runtime. A different message would be needed in this case.
https://github.com/mozilla/rhino/blob/7feaad0e0806e48728e4c7754654b8146db38c30/rhino/src/main/java/org/mozilla/javascript/ScriptableObject.java#L2192-L2200

```js
function NaN() {} // TypeError
```

**Comment by tuchida** [comment]
Fixed by #1530.
